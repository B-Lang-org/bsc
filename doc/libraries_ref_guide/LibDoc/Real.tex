\subsubsection{Real}
\index{Real@\te{Real} (package)}
\label{package-Real}

{\bf Package}

\begin{verbatim}
import Real :: * ;
\end{verbatim}


{\bf Description}

The \te{Real} library package defines functions to operate on and
manipulate real numbers.  Real numbers are numbers with a fractional
component.  They are also of limited precision.  The \te{Real} data
type is described in section \ref{sec-real}.
 

{\bf Constants}

\index{pi@\te{pi} (Real constant)}
\index{$\pi$ (Real constant)}

The constant \te{pi} ($\pi$) is defined.
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{pi}&The value of the constant pi ($\pi$).\\
&\\
\cline{2-2}
&\begin{libverbatim}
 Real pi;
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}



{\bf Trigonometric Functions}



 
The following  trigonometric functions are provided: \te{sin},
\te{cos}, \te{tan}, \te{sinh}, \te{cosh}, \te{tanh}, \te{asin},
\te{acos}, \te{atan}, \te{asinh}, \te{acosh}, \te{atanh}, and \te{atan2}.


\index{sin@\te{sin} (Real function)}
\index[function]{Real!sin}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{sin}&Returns the sine of \te{x}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real sin (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{cos@\te{cos} (Real function)}
\index[function]{Real!cos}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{cos}&Returns the cosine of \te{x}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real cos (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{tan@\te{tan} (Real function)}
\index[function]{Real!tan}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{tan}&Returns the tangent of \te{x}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real tan (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}



\index{sinh@\te{sinh} (Real function)}
\index[function]{Real!sinh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{sinh}&Returns the hyperbolic sine of \te{x}. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real sinh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{cosh@\te{cosh} (Real function)}
\index[function]{Real!cosh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{cosh}&Returns the hyperbolic cosine of \te{x}. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real cosh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{tanh@\te{tanh} (Real function)}
\index[function]{Real!tanh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{tanh}&Returns the hyperbolic tangent of \te{x}. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real tanh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{asinh@\te{asinh} (Real function)}
\index[function]{Real!asinh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{asinh}&Returns the  inverse hyperbolic  sine of \te{x}. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real asinh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{acosh@\te{acosh} (Real function)}
\index[function]{Real!acosh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{acosh}&Returns the inverse hyperbolic cosine of \te{x}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real acosh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{atanh@\te{atanh} (Real function)}
\index[function]{Real!atanh}
\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{atanh}&Returns the inverse hyperbolic  tangent of \te{x}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real atanh (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{atan2@\te{atan2} (Real function)}
\index[function]{Real!atan2}


\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{atan2}&Returns  \te{atan($x/y$)}.  \te{atan2(1,x)} is equivalent to
\te{atan(x)}, but provides more precision when required by the division of
\te{x/y}. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real atan2 (Real y, Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

{\bf Arithmetic Functions}

\index{pow@\te{pow} (Real function)}
\index[function]{Real!pow}


\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{pow}& The element x is raised to the y power.  An alias for
\te{**}. \te{pow(x,y)} = \te{x**y} = $x^y$. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real pow (Real x, Real y);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}


\index{sqrt@\te{sqrt} (Real function)}
\index[function]{Real!sqrt}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{sqrt}& Returns the square root of \te{x}.  Returns an error if \te{x} is negative.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Real sqrt (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}



{\bf Conversion Functions}

The following four functions are used to convert a \te{Real} to an \te{Integer}.

\index{trunc@\te{trunc} (Real function)}
\index[function]{Real!trunc}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{trunc}& Converts a \te{Real} to an \te{Integer} by removing  the fractional
part of \te{x}, which  can be positive or negative.  \te{trunc(1.1) = 1}, \te{trunc(-1.1)= -1}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Integer trunc (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{round@\te{round} (Real function)}
\index[function]{Real!round}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{round}& Converts a \te{Real} to an \te{Integer} by rounding to the nearest
whole number.  \te{.5} rounds up in magnitude. \te{round(1.5) = 2},
\te{round(-1.5)= -2}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Integer round (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{ceil@\te{ceil} (Real function)}
\index[function]{Real!ceil}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{ceil}& Converts a \te{Real} to an \te{Integer} by rounding to the higher
number, regardless of sign.  \te{ceil(1.1) = 2}, \te{ceil(-1.1) = -1}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Integer ceil (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{floor@\te{floor} (Real function)}
\index[function]{Real!floor}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{floor}&Converts a \te{Real} to an \te{Integer} by rounding to the lower
number, regardless of sign.  \te{floor(1.1) = 1}, \te{floor(-1.1) = -2}.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Integer floor (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

There are also two system functions \te{\$realtobits} and
\te{\$bitstoreal}, defined in the Prelude (section \ref{sec-real})
which provide conversion to and from IEEE 64-bit vectors (\te{Bit\#(64)}).

{\bf Introspection Functions}

\index{isInfinite@\te{isInfinite} (Real function)}
\index[function]{Real!isInfinite}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{isInfinite}& Returns  \te{True} if the value of \te{x} is
infinite, \te{False} if \te{x} is finite.\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Bool isInfinite (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{isNegativeZero@\te{isNegativeZero} (Real function)}
\index[function]{Real!isNegativeZero}

\begin{center}
\begin{tabular}{|p{1 in}|p{4.5 in}|}
 \hline
&\\
\te{isNegativeZero}& Returns  \te{True} if the value of \te{x} is negative zero. \\
&\\
\cline{2-2}
&\begin{libverbatim}
function Bool isNegativeZero (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{splitReal@\te{splitReal} (Real function)}
\index[function]{Real!splitReal}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{splitReal}&   Returns a Tuple containing the whole ($n$) and
fractional ($f$)  parts of \te{x} such that $n+f = x$. Both values have
the same sign as \te{x}.  The absolute 
value of the fractional part is guaranteed to be in the range [0,1).\\
&\\
\cline{2-2}
&\begin{libverbatim}
function Tuple2#(Integer, Real) splitReal (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{decodeReal@\te{decodeReal} (Real function)}
\index[function]{Real!decodeReal}

\begin{center}
\begin{tabular}{|p{.8 in}|p{4.7 in}|}
 \hline
&\\
\te{decodeReal}&   Returns a Tuple3 containing the sign, the fraction,
and the exponent of a real number. 
The second part (the first Integer) represents the fractional part as
a signed Integer value.  This can be converted to an \te{Int\#(54)} (52
bits, plus hidden bit, plus the sign bit).  The last value is a signed Integer
representing the exponent, which can be be converted to an \te{Int\#(11)} .
The real number is represented exactly as ($fractional \times
2^{exp}$).   The \te{Bool} represents the sign and
is \te{True} for 
positive and positive zero,  \te{False} for negative and negative
zero. Since the second value is a signed value, the \te{Bool} is
redundant except for zero values.\\ 

&\\
\cline{2-2}
&\begin{libverbatim}
function Tuple3#(Bool, Integer, Integer) decodeReal (Real x);
\end{libverbatim}
\\ \hline
\end{tabular}
\end{center}

\index{realToDigits@\te{realToDigits} (Real function)}
\index[function]{Real!realToDigits}

\begin{center}
\begin{tabular}{|p{1 in}|p{4.5in}|}
\hline
\te{realToDigits}&Deconstructs a real number into its digits.  The
function takes a base and a real number and returns a list of digits
and an exponent (ignoring the sign). In particular, if $x \geq 0$, and
\te{realToDigits(base,x)} returned a list of digits \te{d1, d2, ..., dn} and
an exponent \te{e}, then:
\begin{itemize}
\item \te{n} $\geq$ \te{1}
\item \te{abs(x) = 0.d1d2...dn * (base$^e$)}
\item \te{0} $\leq$ \te{di} $\leq$ \te{base-1}  
\end{itemize}\\
\cline{2-2}
&\begin{libverbatim}
function Tuple2#(List#(Integer), Integer) 
         realToDigits (Integer base, Real r);
\end{libverbatim}
\\
\hline
\end{tabular}
\end{center}



