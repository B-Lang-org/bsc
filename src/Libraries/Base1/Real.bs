package Real(

        -- trigonometric functions
        sin, cos, tan, sinh, cosh, tanh,
        asin, acos, atan, asinh, acosh, atanh,
        atan2,
        -- exponentiation
        pow,
        -- square root
        sqrt,
        -- constants
        pi,
        -- rounding
        trunc, ceil, floor, round,
        -- introspection
        splitReal, decodeReal, realToDigits,
        isInfinite, isNegativeZero

        ) where

-- ---------------
-- trigonometric functions

sin :: Real -> Real
sin = primRealSin

cos :: Real -> Real
cos = primRealCos

tan :: Real -> Real
tan = primRealTan

sinh :: Real -> Real
sinh = primRealSinH

cosh :: Real -> Real
cosh = primRealCosH

tanh :: Real -> Real
tanh = primRealTanH

asin :: Real -> Real
asin = primRealASin

acos :: Real -> Real
acos = primRealACos

atan :: Real -> Real
atan = primRealATan

asinh :: Real -> Real
asinh = primRealASinH

acosh :: Real -> Real
acosh = primRealACosH

atanh :: Real -> Real
atanh = primRealATanH

atan2 :: Real -> Real -> Real
atan2 = primRealATan2

-- ---------------
-- exponentiation

pow :: Real -> Real -> Real
pow = (**)

-- ---------------
-- square root

sqrt :: Real -> Real
sqrt = primRealSqrt

-- ---------------
-- constants

pi :: Real
pi = 3.14159265358979323846

-- ---------------
-- rounding

trunc :: Real -> Integer
trunc = primRealTrunc

ceil :: Real -> Integer
ceil = primRealCeil

floor :: Real -> Integer
floor = primRealFloor

round :: Real -> Integer
round = primRealRound

-- ---------------
-- introspection

splitReal :: Real -> (Integer, Real)
splitReal x = primSplitReal x

decodeReal :: Real -> (Bool, Integer, Integer)
decodeReal x =
    let (b, m, e) = primDecodeReal x
    in  (primChr b, m, e)

realToDigits :: Integer -> Real -> (List Integer, Integer)
realToDigits = primRealToDigits

isInfinite :: Real -> Bool
isInfinite x = primChr (primRealIsInfinite x)

isNegativeZero :: Real -> Bool
isNegativeZero x = primChr (primRealIsNegativeZero x)

-- ---------------

