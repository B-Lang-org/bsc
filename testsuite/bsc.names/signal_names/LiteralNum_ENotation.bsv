// ---------------------------------------------------------------------------

import Real::*;

(* synthesize *)
module sysLiteralNum_ENotation ();

  Reg#(Bool) rg_start <- mkReg(True);

  rule do_disp (rg_start);
    rg_start <= False;
    for (Integer i=0; i<4; i=i+1) begin
      Real r = (pi * (2**fromInteger(i))) / (2**10);
      $display("%d: %f", i, r);
    end
    $finish(0);
  endrule

endmodule

// ---------------------------------------------------------------------------
