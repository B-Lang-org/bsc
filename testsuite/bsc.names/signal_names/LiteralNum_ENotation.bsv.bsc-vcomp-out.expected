checking package dependencies
compiling LiteralNum_ENotation.bsv
code generation for sysLiteralNum_ENotation starts
=== ATS:
APackage sysLiteralNum_ENotation
-- APackage parameters
[]
-- APackage arguments
clock { osc = CLK }
reset { RST_N }
-- APackage wire info
clock info  clock default_clock(CLK, {-inhigh-});
reset info  reset default_reset(RST_N) clocked_by(default_clock);
arg info  [clockarg default_clock;, resetarg default_reset;]
-- APackage clock domains
[(0, [{ osc:  CLK gate:  1'd1 }])]
-- APackage resets
[(0, { wire:  RST_N })]
-- AP state elements
rg_start :: ABSTRACT:  Prelude.VReg = RegN
				      (VModInfo
				       RegN
				       clock _clk__1(CLK, {-unused-});
				       reset _rst__1(RST) clocked_by(_clk__1);
				       [clockarg _clk__1;, resetarg _rst__1;, param width;, param init;]
				       [method (Q_OUT, [reg])read clocked_by (_clk__1) reset_by (_rst__1);,
					method write((D_IN, [reg])) enable ((EN,
									     [])) clocked_by (_clk__1) reset_by (_rst__1);]
				       SchedInfo [read CF read, read SB write, write SBR write] [] [] []
				       [])
				      [clock { osc:  CLK gate:  1'd1 }, reset { wire:  RST_N }, 32'd1, 1'd1]
				      []
				      meth types=[([], Nothing, Just (Bit 1)), ([Bit 1], Just (Bit 1), Nothing)]
				      port types=D_IN -> Prelude.Bool
						 Q_OUT -> Prelude.Bool
-- AP local definitions
rg_start___d1 :: Bit 1;
rg_start___d1  = rg_start.read;
-- IdProp rg_start___d1[IdP_from_rhs]
signed_0___d2 :: Bit 32;
signed_0___d2  = Prelude.$signed 32'd0;
-- IdProp signed_0___d2[IdP_from_rhs,IdP_signed]
signed_1___d4 :: Bit 32;
signed_1___d4  = Prelude.$signed 32'd1;
-- IdProp signed_1___d4[IdP_from_rhs,IdP_signed]
signed_2___d6 :: Bit 32;
signed_2___d6  = Prelude.$signed 32'd2;
-- IdProp signed_2___d6[IdP_from_rhs,IdP_signed]
signed_3___d8 :: Bit 32;
signed_3___d8  = Prelude.$signed 32'd3;
-- IdProp signed_3___d8[IdP_from_rhs,IdP_signed]
_3d0679615757712823eneg3___d3 :: Real ;
_3d0679615757712823eneg3___d3  = 3.0679615757712823e-3;
-- IdProp _3d0679615757712823eneg3___d3[IdP_from_rhs]
_6d135923151542565eneg3___d5 :: Real ;
_6d135923151542565eneg3___d5  = 6.135923151542565e-3;
-- IdProp _6d135923151542565eneg3___d5[IdP_from_rhs]
_1d227184630308513eneg2___d7 :: Real ;
_1d227184630308513eneg2___d7  = 1.227184630308513e-2;
-- IdProp _1d227184630308513eneg2___d7[IdP_from_rhs]
_2d454369260617026eneg2___d9 :: Real ;
_2d454369260617026eneg2___d9  = 2.454369260617026e-2;
-- IdProp _2d454369260617026eneg2___d9[IdP_from_rhs]
-- AP rules
rule RL_do_disp "do_disp":
 when rg_start___d1
  ==> { rg_start.write 1'd0;
	Prelude.$display "%d: %f" signed_0___d2 _3d0679615757712823eneg3___d3;
	Prelude.$display "%d: %f" signed_1___d4 _6d135923151542565eneg3___d5;
	Prelude.$display "%d: %f" signed_2___d6 _1d227184630308513eneg2___d7;
	Prelude.$display "%d: %f" signed_3___d8 _2d454369260617026eneg2___d9;
	Prelude._finish_ 32'd0; }
[]
clock domain = Just (0), resets = [0]
-- AP scheduling pragmas
[]
-- AP interface
-- AP instance comments
-- AP remaining proof obligations
[]

-----

Verilog file created: sysLiteralNum_ENotation.v
All packages are up to date.
